go-controller/pkg/clustermanager/endpointslicemirror/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,428 @@
+func isManagedByDefault(endpointSlice *v1.EndpointSlice) bool {
+	return types.EndpointSliceDefaultControllerName == endpointSlice.Labels[v1.LabelManagedBy]
+}
+
+// getPodIP retrieves the IP address of a specified Pod within a given namespace and network.
+// If the pod is host networked it returns default pod IP from the status ignoring the network.
+// Otherwise, it unmarshals the Pod's network annotation, and matches the IP from the provided network.
+func (c *Controller) getPodIP(name, namespace, network string, isIPv6 bool) (string, error) {
+	var podIP string
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		return "", err
+	}
+
+	if pod.Spec.HostNetwork {
+		podIPs, err := util.DefaultNetworkPodIPs(pod)
+		if err != nil {
+			return "", err
+		}
+
+		// MatchFirstIPFamily returns an error if no IP was found
+		ipAddr, err := util.MatchFirstIPFamily(isIPv6, podIPs)
+		if err != nil {
+			return "", err
+		}
+
+		podIP = ipAddr.String()
+	} else {
+		net, err := util.UnmarshalPodAnnotation(pod.Annotations, network)
+		if err != nil {
+			return "", err
+		}
+
+		// MatchFirstIPFamily returns an error if no IP was found
+		ipNet, err := util.MatchFirstIPNetFamily(isIPv6, net.IPs)
+		if err != nil {
+			return "", err
+		}
+		podIP = ipNet.IP.String()
+	}
+
+	return podIP, nil
+}
+
+// mirroredEndpointSlice creates or updates a mirrored EndpointSlice based on the provided defaultEndpointSlice.
+// The mirrored EndpointSlice will have custom labels set and will be managed by the current controller.
+func (c *Controller) mirroredEndpointSlice(mirroredEndpointSlice, defaultEndpointSlice *v1.EndpointSlice, network string) *v1.EndpointSlice {
+	var currentMirror *v1.EndpointSlice
+	if mirroredEndpointSlice != nil {
+		currentMirror = mirroredEndpointSlice.DeepCopy()
+	}
+	if currentMirror == nil {
+		currentMirror = &v1.EndpointSlice{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:       defaultEndpointSlice.Namespace,
+				OwnerReferences: defaultEndpointSlice.OwnerReferences,
+			},
+			AddressType: defaultEndpointSlice.AddressType,
+			Ports:       defaultEndpointSlice.Ports,
+		}
+	}
+	if currentMirror.Labels == nil {
+		currentMirror.Labels = map[string]string{}
+	}
+	// set the custom labels, generateName and reset the endpoints
+	currentMirror.Labels[v1.LabelManagedBy] = c.name
+	currentMirror.Labels[LabelSourceEndpointSlice] = defaultEndpointSlice.Name
+	currentMirror.Labels[LabelSourceEndpointSliceVersion] = defaultEndpointSlice.ResourceVersion
+	currentMirror.Labels[types.LabelUserDefinedEndpointSliceNetwork] = network
+	currentMirror.Labels[types.LabelUserDefinedServiceName] = defaultEndpointSlice.Labels[v1.LabelServiceName]
+
+	origGenName := defaultEndpointSlice.GenerateName
+	if len(defaultEndpointSlice.GenerateName) == 0 {
+		origGenName = defaultEndpointSlice.Name
+	}
+	currentMirror.GenerateName = getGenerateName(origGenName, network)
```

Comment:
According to the k8s code it looks like GenerateName and Name should not both be set, and Name should be unique. Earlier in this function you do a deep copy, but looking at that code it doesn't look to copy Name. So is that why you are always setting GenerateName here?

If yes, then it looks like you are leveraging that to do a check in the sync function:
```
if !reflect.DeepEqual(currentMirror, mirroredEndpointSlice) {
		if currentMirror.Name == "" {
```
If name is empty, then you know it doesn't exist yet in kapi so it should be a create. Am I understanding correctly?

Reply:
Good spot!
DeepCopy copies the name but setting `GenerateName` when `Name` is set doesn't make sense.
It worked because the Update calls ignored `GenerateName` but I fixed it for correctness.